x/ibc/07-tendermint/misbehaviour.go

```diff
@@ -94,47 +76,55 @@ func checkMisbehaviour(
 		consensusParams.Evidence != nil &&
 		ageDuration > consensusParams.Evidence.MaxAgeDuration &&
 		ageBlocks > uint64(consensusParams.Evidence.MaxAgeNumBlocks) {
-		return sdkerrors.Wrapf(clienttypes.ErrInvalidEvidence,
+		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidEvidence,
 			"age duration (%s) and age blocks (%d) are greater than max consensus params for duration (%s) and block (%d)",
 			ageDuration, ageBlocks, consensusParams.Evidence.MaxAgeDuration, consensusParams.Evidence.MaxAgeNumBlocks,
 		)
 	}
 
-	// check if provided height matches the headers' height
-	if height > uint64(evidence.GetHeight()) {
-		return sdkerrors.Wrapf(
-			sdkerrors.ErrInvalidHeight,
-			"height > evidence header height (%d > %d)", height, evidence.GetHeight(),
-		)
+	// Check the validity of the two conflicting headers against their respective
+	// trusted consensus states
+	if err := checkMisbehavingHeader(
+		tmClientState, tmConsensusState1, tmEvidence.Header1, currentTimestamp,
+	); err != nil {
+		return nil, sdkerrors.Wrap(err, "verifying Header1 in Evidence failed")
+	}
+	if err := checkMisbehavingHeader(
+		tmClientState, tmConsensusState2, tmEvidence.Header2, currentTimestamp,
+	); err != nil {
+		return nil, sdkerrors.Wrap(err, "verifying Header2 in Evidence failed")
 	}
 
-	// NOTE: header height and commitment root assertions are checked with the
-	// evidence and msg ValidateBasic functions at the AnteHandler level.
+	tmClientState.FrozenHeight = uint64(tmEvidence.GetHeight())
+	return tmClientState, nil
+}
+
+// checkMisbehavingHeader checks that a Header in Misbehaviour is valid evidence given
+// a trusted ConsensusState
+func checkMisbehavingHeader(
+	clientState *types.ClientState, consState *types.ConsensusState, header types.Header, currentTimestamp time.Time,
+) error {
+	// check TrustedFields against ConsensusState
```

Comment:
```suggestion
	// check the trusted fields for the header against ConsensusState
```
we should only use camelCase for names that exist in the code. Is `TrustedFields` defined somewhere?